libz-rs-sys/src/gz.rs

```diff
@@ -753,6 +812,188 @@ pub unsafe extern "C-unwind" fn gzdirect(file: gzFile) -> c_int {
+// or an error. `state.err` must be consulted to determine which.
+//
+// # Safety
+//
+// * `state` must have been properly initialized, e.g. by [`gzopen_help`].
+// * `buf` must point to at least `len` bytes of writable memory.
+unsafe fn gz_read(state: &mut GzState, mut buf: *mut u8, mut len: usize) -> usize {
+    if len == 0 {
+        return 0;
+    }
+
+    /* FIXME: Uncomment this when seek support is added:
+
+    // process a skip request
+    if state.seek {
+        state.seek = false;
+        if gz_skip(state, state.skip) == -1 {
+            return 0;
+        }
+    }
+     */
+
+    // Loop until we get enough bytes or reach the end of the file.
+    let mut got = 0;
+    let mut count = 0;
+    loop {
+        count += 1;
+        assert_ne!(count, 100);
+
+        // Set n to the maximum amount of len that fits in an unsigned int.
+        let mut n = cmp::min(len, c_uint::MAX as usize);
+
+        // First just try copying data from the output buffer. Note: The output
+        // buffer contains bytes that have been decompressed by `state.stream` and
+        // are waiting to be consumed - or, in direct mode, it contains bytes read
+        // directly from the underlying file descriptor.
+        if state.have != 0 {
+            n = cmp::min(n, state.have as usize);
+            // Safety:
+            // * n <= state.have, and there are `state.have` readable bytes starting
+            //   at `state.next`.
+            // * n <= len, and the caller is responsible for ensuring that `buf`
+            //   points to at least `len` writable bytes.
+            // * `state.next` points into an internal buffer not visible outside the
+            //   `GzState`, and `buf` is supplied by the caller, so the source and
+            //   destination are guaranteed not to overlap.
+            unsafe { ptr::copy_nonoverlapping(state.next, buf, n) };
+            state.next = unsafe { state.next.add(n) };
+            state.have -= n as c_uint;
+        } else if state.eof && state.stream.avail_in == 0 {
+            // The output buffer is empty, and we're at the end of the file.
+            state.past = true; // Tried to read past end
+            break;
+        } else if state.how == How::Look || n < state.in_size * 2 {
+            // For small len or a new stream, load more data from the file descriptor into
+            // the output buffer. Note: If we haven't scanned the file header yet, gz_fetch
+            // will read the header and determine whether to use decompression or direct read.
+            if unsafe { gz_fetch(state) }.is_err() {
+                return 0;
+            }
+            assert_ne!(state.have, 0);
+            continue; // Now that we've tried reading, we can try to copy from the output buffer.
+                      // The copy above assures that we will leave with space in the
+                      // output buffer, allowing at least one gzungetc() to succeed.
```

Comment:
this reads nicer if you put the comment above the `continue` (and maybe add an empty line after the `assert_ne`